lib/features/bottom_nav_bar/bottom_nav_bar_controller.dart

```diff
@@ -1,29 +1,24 @@
 import 'package:riverpod_annotation/riverpod_annotation.dart';
 
+import '../../config.dart';
+import '../navigator/utils/selected_tab_observer.dart';
 import 'nav_bar_config.dart';
-import 'nested_navigator.dart';
 
 part 'bottom_nav_bar_controller.g.dart';
 
-@riverpod
+@Riverpod(keepAlive: true)
 class BottomNavBarController extends _$BottomNavBarController {
+  late final SelectedTabObserver selectedTabObserver =
+      SelectedTabObserver(_onTabChanged);
+
   @override
   NavBarEnum build() {
-    return NavBarEnum.home;
+    return MyAppConfig.initialTab;
   }
 
-  void onIndexChanged(int index) {
-    if (index == state.index) {
-      return; // prevents potencial popping of a nested view, when clicked on a current tab
+  void _onTabChanged(NavBarEnum tab) {
+    if (state != tab) {
+      state = tab;
```

Comment:
Do you think the if statement is needed here? Is the performance significantly better that way? I'm just wondering, because with plain variables, according to my best knowledge, it doesn't.

Reply:
Actually, as far I know, you are right :P. But I believe I had some problems, that's why I've added this `if` - but maybe it was due to different reasons. I'll defienetly check this out and test again

Reply:
So first things, you're right, riverpod checks if the state has changed and rebuilds it only when it did. Here's the riverpod source code
<img width="628" alt="Zrzut ekranu 2024-06-5 o 22 06 02" src="https://github.com/Solvro/mobile-topwr/assets/28555148/557658e6-e40a-43e5-9273-73fa4cac5b28">

Reply:
But I obviously knew this. But for a still unknown reasons for me, somehow due to usage our `RouteObserver<PageRoute<dynamic>>` it gives an error - but only if we include this enum in the args of the initial route. If we have null args in the initial route, it doesn't break when we jump to next ones. The problem exists only in first render of the app
![Screenshot_1717618165](https://github.com/Solvro/mobile-topwr/assets/28555148/28da5751-693f-4613-9c8c-23bae38a404f)

Reply:
The reason of the error is that it somehow is triggering rebuild of the ui in unfinished state of rendering - which is forbidden due to risk of never ending render (when one render causes re-render before it's ended = never ending single render which trapped in infinite loop) - that's why we see the error. 

But it only breaks during first render and for unknown reasons the `if-check` helps. Which suggests that riverpod's internal check is somehow broken during the initial render. 

Also, for you to know for the future, when you see error like that, the easiest always working fix is to make the change-trigger asonchrynous like that:
<img width="470" alt="Zrzut ekranu 2024-06-5 o 22 12 45" src="https://github.com/Solvro/mobile-topwr/assets/28555148/e8689a15-aa7e-4ce1-b461-ec8d91404093">
or like that:
<img width="664" alt="Zrzut ekranu 2024-06-5 o 22 13 00" src="https://github.com/Solvro/mobile-topwr/assets/28555148/40650a54-7f56-4674-9731-9d4185ed0ef2">
it can be in this place as well:
<img width="689" alt="Zrzut ekranu 2024-06-5 o 22 13 33" src="https://github.com/Solvro/mobile-topwr/assets/28555148/de702dd7-602a-4f4c-a7ae-b7add8d12a98">

but generally speaking, in most cases, it may suggest a wrong usage of something, but the workaround almost always works, but should be used with some moderation. It's better to look for other solution if possible. 

Yeah, but somehow this if also fix this:
<img width="447" alt="Zrzut ekranu 2024-06-5 o 22 15 44" src="https://github.com/Solvro/mobile-topwr/assets/28555148/674c5e4d-5779-4c26-9a2d-bd55ec174e02">

I'll look into this tomorrow more, maybe I'm doing something wrong but this is kinda weird for me.

Reply:
Thank you for explaining all of this. I hope we could find a solution here, maybe the next article is coming 😎 

 For sure at some point in the future I will need to take advantage of advices you gave me about errors like that, thanks for this! 

I'm curious about possible fix but not to block the commit we could leave it that way I guess.

Good job!

Reply:
Yeah, maybe it is😎 But rn the university is killing me, I'm already half dead. When doing quick tests, it seems like the riverpod doesn't call updateShouldNotify in this first build? idk it's werid
<img width="809" alt="Zrzut ekranu 2024-06-5 o 22 45 13" src="https://github.com/Solvro/mobile-topwr/assets/28555148/032f9203-d185-419b-b10c-2affe9a7cb0c">

Reply:
I'll leave it with `if` and we'll look into that after I pass this term (hopefully)